Mage.Sets/src/mage/cards/k/KaghaShadowArchdruid.java

```diff
@@ -0,0 +1,154 @@
+    }
+
+    private KaghaShadowArchdruidEffect(final KaghaShadowArchdruidEffect effect) {
+        super(effect);
+    }
+
+    @Override
+    public KaghaShadowArchdruidEffect copy() {
+        return new KaghaShadowArchdruidEffect(this);
+    }
+
+    @Override
+    public boolean apply(Game game, Ability source) {
+        return true;
+    }
+
+    @Override
+    public boolean applies(UUID objectId, Ability source, UUID affectedControllerId, Game game) {
+        if (source.isControlledBy(affectedControllerId)
+                && Zone.GRAVEYARD.equals(game.getState().getZone(objectId))
+                && game.isActivePlayer(affectedControllerId)) {
+            Card card = game.getCard(objectId);
+            Permanent sourceObject = source.getSourcePermanentIfItStillExists(game);
+            KaghaShadowArchdruidWatcher watcher = game.getState().getWatcher(KaghaShadowArchdruidWatcher.class);
+            if (card != null && sourceObject != null 
+                    && card.isOwnedBy(affectedControllerId)
+                    && watcher != null
+                    && watcher.checkCard(card, game)) {
+                return watcher.abilityNotUsed(new MageObjectReference(sourceObject, game));
+            }
+        }
+        return false;
+    }
+}
+
+class KaghaShadowArchdruidWatcher extends Watcher {
+
+    private final Set<MageObjectReference> usedFrom = new HashSet<>();
+    private final Set<MageObjectReference> morSet = new HashSet<>();
+
+    KaghaShadowArchdruidWatcher() {
+        super(WatcherScope.GAME);
+    }
+
+    @Override
+    public void watch(GameEvent event, Game game) {
+        if ((GameEvent.EventType.SPELL_CAST.equals(event.getType()) || GameEvent.EventType.LAND_PLAYED.equals(event.getType()))
+                && event.hasApprovingIdentifier(MageIdentifier.KaghaShadowArchdruidWatcher)) {
+            usedFrom.add(event.getAdditionalReference().getApprovingMageObjectReference());
+        }
+        if (event.getType() == GameEvent.EventType.ZONE_CHANGE) {
+            ZoneChangeEvent zEvent = (ZoneChangeEvent) event;
+            if (zEvent.getFromZone() == Zone.LIBRARY && zEvent.getToZone() == Zone.GRAVEYARD) {
+                morSet.add(new MageObjectReference(zEvent.getTargetId(), game));
+            }
+        }
+    }
+
+    @Override
+    public void reset() {
+        super.reset();
+        usedFrom.clear();
+        morSet.clear();
+    }
+
+    boolean abilityNotUsed(MageObjectReference mor) {
+        return !usedFrom.contains(mor);
+    }
+
+    boolean checkCard(Card card, Game game) {
+        return card.isPermanent(game) && game
+                .getState()
+                .getWatcher(KaghaShadowArchdruidWatcher.class)
+                .morSet
+                .stream()
+                .anyMatch(mor -> mor.refersTo(card, game));
```

Comment:
You can replace `game.getState().getWatcher(..)` by `this`, as this is the watcher's method.